feat(platform-server): allow shimming the global env sooner

gkalpak opened the pull request:
`@angular/platform-server` provides the foundation for rendering an Angular app on the server. In order to achieve that, it uses a server-side DOM implementation (currently [domino][1]).

For rendering on the server to work as closely as possible to running the app on the browser, we need to make DOM globals (such as `Element`, `HTMLElement`, etc.), which are normally provided by the browser, available as globals on the server as well.

Currently, `@angular/platform-server` achieves this by extending the `global` object with the DOM implementation provided by `domino`. This assignment happens in the [setDomTypes()][2] function, which is [called in a `PLATFORM_INITIALIZER`][3]. While this works in most cases, there are some scenarios where the DOM globals are needed sooner (i.e. before initializing the platform). See, for example, #24551 and #39950 for more details on such issues.

This commit provides a way to solve this problem by exposing an side-effect-ful entry-point (`@angular/platform-server/shims`), that shims the `global` object with DOM globals. People will be able to import this entry-point in their server-rendered apps before bootstrapping the app (for example, in their `main.server.ts` file).
(See also [#39950 (comment)][4].)

In a future update, the [`universal` schematics][5] will include such an import by default in newly generated projects.

##
Closes #39950.

Related to angular/angular-cli#19869.

[1]: https://www.npmjs.com/package/domino
[2]: https://github.com/angular/angular/blob/0fc8466f1be392917e0c/packages/platform-server/src/domino_adapter.ts#L17-L21
[3]: https://github.com/angular/angular/blob/0fc8466f1be392917e0c/packages/platform-server/src/server.ts#L33
[4]: https://github.com/angular/angular/issues/39950#issuecomment-747598403
[5]: https://github.com/angular/angular-cli/blob/cc51432661eb4ab4b6a3/packages/schematics/angular/universal

mary-poppins commented:
You can preview 5aae56d at https://pr40559-5aae56d.ngbuilds.io/.

gkalpak commented on packages/platform-server/package.json:
```diff
@@ -25,7 +25,7 @@
     "packageGroup": "NG_UPDATE_PACKAGE_GROUP"
   },
-  "sideEffects": false,
+  "sideEffects": true,
```
I was hoping that it would be sufficient to set `sideEffects: true` in `platform-server/shims/package.json`, but the import seems to be dropped if I don't specify it here as well 😞

alan-agius4 replied:
The `sideEffects` flag only work in the primary entrypoint.

Consider, listing which entry-points are non side-effect free. Example:
```js
"sideEffects": [
  './bundles/shims.umd.js',
  './bundles/shims.umd.min.js',
  './fesm2015/shims.js',
  './esm2015/shims/src/*.js',
  './__ivy_ngcc__/shims.umd.min.js',
  './__ivy_ngcc__/fesm2015/shims.js',
  './__ivy_ngcc__/bundles/shims.js',
  './__ivy_ngcc__/esm2015/shims/src/*.js',
] 
```

More info https://webpack.js.org/guides/tree-shaking/#mark-the-file-as-side-effect-free

alan-agius4 commented on integration/cli-elements-universal/package.json:
```diff
@@ -0,0 +1,62 @@
+{
```
I don't think we should add dependencies on nguniversal. 

We can test Elements on the platform server without nguniversal.

gkalpak replied:
You had proposed that [before (non-public link)][1] and I remember trying to do this for #39994 (which is where I cherry-picked this commit from) and it didn't work.

I don't recall what the problem was (:grin:), but I suspect I couldn't find a way to test the fix. For example, running `ng run <my-app>:server:production` does not throw an error even without the fix (because the error is a runtime one).

Any ideas on how we could test this?

[1]: https://angular-team.slack.com/archives/C0DH5UDN1/p1607102616074200

alan-agius4 replied:
`ng run <my-app>:server:production` doesn't throw the error because it just build the server bundle. The easiest way would be to actual generate an app-shell. Under the hood is pretty similar and you will have the same behaviour during runtime. 

The test run will also be faster because you don't have the overhead of the prerender builder, and additional dependencies needed by nguniversal.

```
ng g app-shell
ng run my-app:app-shell <- This builds the browser & server bundles and runs the the generated server bundle to generate the app-shell.
```

gkalpak replied:
Aha! That worked :tada: Thx :heart:

alan-agius4 commented on packages/platform-server/shims/package.json:
```diff
@@ -0,0 +1,4 @@
+{
+  "name": "@angular/platform-server/shims",
+  "sideEffects": true
```
`sideEffects` flag can only be defined in primary entry-points.

This has no effect here.

```suggestion
  "sideEffects": true
```

gkalpak replied:
It seems that other secondary entry-points have this (for example, [platform-browser/animations][1], [router/upgrade][2]), so I suggest we keep it. Besides consistency with other entry-points, I think it is a good indication (to humans) that this entry-point if side-effect-ful.

[1]: https://github.com/angular/angular/blob/956f75f0689c73c8ea6798c8dec5d7ee38948bc7/packages/platform-browser/animations/package.json
[2]: https://github.com/angular/angular/blob/956f75f0689c73c8ea6798c8dec5d7ee38948bc7/packages/router/upgrade/package.json

alan-agius4 replied:
Indeed, although it might give the wrong impression. That this does field does something. I think humans should check the file that the tools use to determine which entry-point is side-effect free or not.

Anyways, leaving it is not a problem. It just can cause confusion. So, feel free to do whatever you feel.

mary-poppins commented:
You can preview 6340fd7 at https://pr40559-6340fd7.ngbuilds.io/.

mary-poppins commented:
You can preview f601f3f at https://pr40559-f601f3f.ngbuilds.io/.

mary-poppins commented:
You can preview e8626aa at https://pr40559-e8626aa.ngbuilds.io/.

gkalpak commented:
I have updated the example to not use `@nguniversal` dependencies. PTAL, @alan-agius4.

mary-poppins commented:
You can preview 55b3a2e at https://pr40559-55b3a2e.ngbuilds.io/.

alan-agius4 reviewed (approved):
This does LGTM.

That being said, I think prior to getting this in and committing to adding a side-effect secondary entry-point `@angular/platform-server/shims`.  I think we should decide on the future of other shims such as in https://github.com/angular/universal/pull/1922 and if maybe we want to add some of those shims as part of the platform server.

One of the requirements though for those shims to work properly, is that those shims, will need to be invoked and passed in the template.

//cc @mgechev & @CaerusKaru

CaerusKaru reviewed (approved):
LGTM, though I agree with Alan's sentiment that we may want to sync with @SanderElias and ensure that what he wants to do aligns with this as well. There's a good chance we can get this in before that though, too.

mary-poppins commented:
You can preview 81c0521 at https://pr40559-81c0521.ngbuilds.io/.

kyliau reviewed (approved):
LGTM. 
For context, @alan-agius4 pointed out that explicit call by users to override `globals` is not ergonomic, because the call needs to be invoked prior to any imports that require the shim. (It'd have avoided the use of side-effecty import though).

AndrewKushnir reviewed (approved):
Reviewed-for: public-api

Note the `public-api` group review was triggered by adding an empty
`goldens/public-api/platform-server/shims/shims.d.ts` in this PR. We should probably avoid generating empty golden files at all, but this is outside of the scope of this PR.

petebacondarwin commented on packages/platform-server/package.json:
```diff
@@ -25,7 +25,15 @@
+    "./bundles/platform-server-shims.umd.min.js",
+    "./esm2015/shims/src/shims.js",
+    "./fesm2015/shims.js",
+    "./__ivy_ngcc__/bundles/platform-server-shims.umd.js",
```
I guess that `./__ivy_ngcc__/bundles/platform-server-shims.umd.min.js` is never created by ngcc?

gkalpak replied:
Exactly. There is not `package.json` property pointing to the minified UMD file (and even if there were, ngcc would not be able to process is correctly 😅).

petebacondarwin reviewed (commented):
In `@angular/localize` the side-effectful import is called `init` (e.g. `@angular/localize/init`).
Perhaps we should maintain that convention here, especially since the internal function is even called `init()`!

Reviewed-for: public-api

alan-agius4 commented:
I like the `init` entrypoint name

jelbourn commented:
How likely is it that someone might want to provide an alternate set of shims? On the `init` vs `shims` front, I could still seeing calling this `shims` if there's not any other type of side-effectful initialization that would ever need to happen. 

Also- should we be worried about code formatters re-ordering imports?

jelbourn commented on integration/cli-elements-universal/README.md:
```diff
@@ -0,0 +1,28 @@
+# CliElementsUniversal
+
+This project was generated with [Angular CLI](https://github.com/angular/angular-cli) version 11.0.0-rc.1.
+Support for Angular Elements was added with `ng add @angular/elements` and for SSR with `ng add @nguniversal/express-engine`.
+
+## Development server
+
+Run `ng serve` for a dev server. Navigate to `http://localhost:4200/`. The app will automatically reload if you change any of the source files.
+
+## Code scaffolding
+
+Run `ng generate component component-name` to generate a new component. You can also use `ng generate directive|pipe|service|class|guard|interface|enum|module`.
+
+## Build
+
+Run `ng build` to build the project. The build artifacts will be stored in the `dist/` directory. Use the `--prod` flag for a production build.
+
+## Running unit tests
+
+Run `ng test` to execute the unit tests via [Karma](https://karma-runner.github.io).
+
+## Running end-to-end tests
+
+Run `ng e2e` to execute the end-to-end tests via [Protractor](http://www.protractortest.org/).
+
+## Further help
+
+To get more help on the Angular CLI use `ng help` or go check out the [Angular CLI Overview and Command Reference](https://angular.io/cli) page.
```
Replace this contents of this readme with a description of what the integration test captures?

atscott reviewed (approved):
Reviewed-for: public-api

gkalpak commented:
> How likely is it that someone might want to provide an alternate set of shims?

Not very likely (but not impossible either). I mean, I am sure people might want to provide more shims on top of the domino ones, but _instead_ of the domino ones.

In any case, that should be easy to do: they simply won't import `@angular/platform-server/shims`.

(Also, note that we are already providing the shims at a later stage (once the app bootstraps). This PR only makes it possible to provide the shims earlier in the process).

##
> On the init vs shims front, I could still seeing calling this shims if there's not any other type of side-effectful initialization that would ever need to happen. Consistency is also good, though.

I guess it is hard to tell whether we might need more initialization that just shims in the future. (I don't see it as very likely, but we can't rule it out either.)

Since most people seem to be pro `init`, I am going to change it to that.

##
> Also- should we be worried about code formatters re-ordering imports?

Yes, that would be a problem (although it is arguably not a great idea to re-order imports, since they _can_ have side effects). Note, we had reports of issues due to re-ordering other side-effect-ful imports (e.g. `zone.js/dist/zone-testing`): #40305

There is not much we can do, I guess. In angular/angular-cli#19869, which makes adding this import part of the `universal` schematic, I have added a note above the import explaining that it should come before other imports that rely on DOM built-ins: https://github.com/angular/angular-cli/pull/19869/files#diff-ea3f134f2268461f0fbee755a6bc4340bf729674e338317903484027e4fec71fR4

mary-poppins commented:
You can preview 5392175 at https://pr40559-5392175.ngbuilds.io/.

gkalpak commented:
Comments addressed (the main change is renaming the entry-point from `shims` to `init`).

jessicajaniuk reviewed (approved):
LGTM

reviewed-for: public-api

jelbourn reviewed (approved):
LGTM

Reviewed-for: public-api

gkalpak committed:
build(platform-server): add `domino` to the list of dependencies

The `platform-server` package currently depends on the [domino][1]
package. This commit adds `domino` to the list of dependencies for the
`platform-server` `ng_module` target.

[1]: https://www.npmjs.com/package/domino

gkalpak committed:
ci: add integration test for Angular Elemens with SSR

This commit adds an integration test that uses `@angular/elements` with
`@angular/platform-server` in order to highlight a current
incompatibility. The issue will be fixed in a subsequent commit.

gkalpak committed:
fixup! ci: add integration test for Angular Elemens with SSR

gkalpak committed:
feat(platform-server): allow shimming the global env sooner

`@angular/platform-server` provides the foundation for rendering an
Angular app on the server. In order to achieve that, it uses a
server-side DOM implementation (currently [domino][1]).

For rendering on the server to work as closely as possible to running
the app on the browser, we need to make DOM globals (such as `Element`,
`HTMLElement`, etc.), which are normally provided by the browser,
available as globals on the server as well.

Currently, `@angular/platform-server` achieves this by extending the
`global` object with the DOM implementation provided by `domino`. This
assignment happens in the [setDomTypes()][2] function, which is
[called in a `PLATFORM_INITIALIZER`][3]. While this works in most cases,
there are some scenarios where the DOM globals are needed sooner (i.e.
before initializing the platform). See, for example, #24551 and #39950
for more details on such issues.

This commit provides a way to solve this problem by exposing a
side-effect-ful entry-point (`@angular/platform-server/init`), that
shims the `global` object with DOM globals. People will be able to
import this entry-point in their server-rendered apps before
bootstrapping the app (for example, in their `main.server.ts` file).
(See also [#39950 (comment)][4].)

In a future update, the [`universal` schematics][5] will include such an
import by default in newly generated projects.

[1]: https://www.npmjs.com/package/domino
[2]: https://github.com/angular/angular/blob/0fc8466f1be392917e0c/packages/platform-server/src/domino_adapter.ts#L17-L21
[3]: https://github.com/angular/angular/blob/0fc8466f1be392917e0c/packages/platform-server/src/server.ts#L33
[4]: https://github.com/angular/angular/issues/39950#issuecomment-747598403
[5]: https://github.com/angular/angular-cli/blob/cc51432661eb4ab4b6a3/packages/schematics/angular/universal

gkalpak committed:
fixup! feat(platform-server): allow shimming the global env sooner

gkalpak committed:
fixup! feat(platform-server): allow shimming the global env sooner

mary-poppins commented:
You can preview a67a720 at https://pr40559-a67a720.ngbuilds.io/.
